### ISSUE TITLE
Trust user-added CAs on Android

### ISSUE THREAD

**Issue by gnprice** [zulip/zulip-flutter#461]
This is a configuration change that was requested a few times over the years for zulip-mobile, as https://github.com/zulip/zulip-mobile/issues/3312, before we got a contribution implementing it as https://github.com/zulip/zulip-mobile/pull/5493 . We should do the same thing here.

The implementation for zulip-mobile (see https://github.com/zulip/zulip-mobile/commit/85c3a716a7bbf00c936beaf34bc84a0e6bf5c885) is entirely at the Android metadata layer, so we should be able to pretty much just copy-paste that.

**Comment by gnprice** [comment]
This feature is in [the new beta release v0.0.9](https://chat.zulip.org/#narrow/stream/1-announce/topic/mobile.20beta/near/1716421). @shrizza please give it a try — I'd be glad to have confirmation of whether it works for you.

**Comment by gnprice** [comment]
Seems like that version doesn't yet work — but there is hope that a small further tweak to the manifest will make it work. Details in chat:
https://chat.zulip.org/#narrow/stream/48-mobile/topic/flutter.3A.20user-added.20certs/near/1716835

**Comment by gnprice** [comment]
Rereading the thread, it looks like the next step is to make that small further tweak to our manifest, and then put that in the next beta release so @shrizza can try it out. I think we have another beta release (v0.0.15) coming up this week, so I'll aim to include this there.

(The last round of this was just before I went travelling for part of a week and then got covid; I guess by the time I was back at full speed, this had fallen off my radar.)

**Comment by shrizza** [comment]
I appreciate the proactive follow-up, and look forward to the next release.

**Comment by gnprice** [comment]
Well, unfortunately it looks like I'm going to have to revert the take-2 fix I made as #671.

After that PR, in a release build on Android, the app crashes at startup with SIGSEGV and a stack trace like:
```
#00 pc 00000000006b39e8  /apex/com.android.art/lib64/libart.so (art::JNI<false>::GetStringChars(_JNIEnv*, _jstring*, unsigned char*)+232) (BuildId: 7ece79c15d80914c83e60c9e93ac1684)
#01 pc 00000000000ddbd8  /system/lib64/libandroid_runtime.so (android::android_content_XmlBlock_nativeGetAttributeIndex(_JNIEnv*, _jobject*, long, _jstring*, _jstring*)+72) (BuildId: 2c0c526c96b2f0f37ce26c968792b5fe)
#02 pc 0000000000331474  /data/misc/apexdata/com.android.art/dalvik-cache/arm64/boot.oat (art_jni_trampoline+116)
#03 pc 00000000008bc6d4  /data/misc/apexdata/com.android.art/dalvik-cache/arm64/boot.oat (android.content.res.XmlBlock$Parser.getAttributeBooleanValue+84)
#04 pc 00000000005ba754  /apex/com.android.art/lib64/libart.so (nterp_helper+7636) (BuildId: 7ece79c15d80914c83e60c9e93ac1684)
#05 pc 000000000021525e  /data/app/~~OdBMxcLjRbZre3jUFwqhtA==/com.zulip.flutter-Ivlz_Plsw-9nCNlt4y7GhA==/base.apk (y4.a.c+58)
#06 pc 000000000033b680  /apex/com.android.art/lib64/libart.so (art_quick_invoke_static_stub+640) (BuildId: 7ece79c15d80914c83e60c9e93ac1684)
#07 pc 0000000000511714  /apex/com.android.art/lib64/libart.so (bool art::interpreter::DoCall<false>(art::ArtMethod*, art::Thread*, art::ShadowFrame&, art::Instruction const*, unsigned short, bool, art::JValue*)+2364) (BuildId: 7ece79c15d80914c83e60c9e93ac1684)
#08 pc 000000000049136c  /apex/com.android.art/lib64/libart.so (void art::interpreter::ExecuteSwitchImplCpp<false>(art::interpreter::SwitchImplContext*)+1892) (BuildId: 7ece79c15d80914c83e60c9e93ac1684)
#09 pc 00000000003545d8  /apex/com.android.art/lib64/libart.so (ExecuteSwitchImplAsm+8) (BuildId: 7ece79c15d80914c83e60c9e93ac1684)
#10 pc 0000000000215534  /data/app/~~OdBMxcLjRbZre3jUFwqhtA==/com.zulip.flutter-Ivlz_Plsw-9nCNlt4y7GhA==/base.apk (y4.a.e+0)
#11 pc 000000000036e6ec  /apex/com.android.art/lib64/libart.so (art::interpreter::Execute(art::Thread*, art::CodeItemDataAccessor const&, art::ShadowFrame&, art::JValue, bool, bool) (.__uniq.112435418011751916792819755956732575238.llvm.9545667076320299271)+232) (BuildId: 7ece79c15d80914c83e60c9e93ac1684)
#12 pc 000000000036dfe4  /apex/com.android.art/lib64/libart.so (artQuickToInterpreterBridge+964) (BuildId: 7ece79c15d80914c83e60c9e93ac1684)
#13 pc 0000000000351f68  /apex/com.android.art/lib64/libart.so (art_quick_to_interpreter_bridge+88) (BuildId: 7ece79c15d80914c83e60c9e93ac1684)
#14 pc 00000000005b8a18  /apex/com.android.art/lib64/libart.so (nterp_helper+152) (BuildId: 7ece79c15d80914c83e60c9e93ac1684)
#15 pc 00000000002161a6  /data/app/~~OdBMxcLjRbZre3jUFwqhtA==/com.zulip.flutter-Ivlz_Plsw-9nCNlt4y7GhA==/base.apk (y4.f.t+66)
#16 pc 00000000005b98d4  /apex/com.android.art/lib64/libart.so (nterp_helper+3924) (BuildId: 7ece79c15d80914c83e60c9e93ac1684)
#17 pc 000000000021614a  /data/app/~~OdBMxcLjRbZre3jUFwqhtA==/com.zulip.flutter-Ivlz_Plsw-9nCNlt4y7GhA==/base.apk (y4.f.s+10)
#18 pc 00000000005b98d4  /apex/com.android.art/lib64/libart.so (nterp_helper+3924) (BuildId: 7ece79c15d80914c83e60c9e93ac1684)
#19 pc 000000000017f720  /data/app/~~OdBMxcLjRbZre3jUFwqhtA==/com.zulip.flutter-Ivlz_Plsw-9nCNlt4y7GhA==/base.apk (io.flutter.embedding.engine.d.<init>+56)
#20 pc 00000000005b98d4  /apex/com.android.art/lib64/libart.so (nterp_helper+3924) (BuildId: 7ece79c15d80914c83e60c9e93ac1684)
#21 pc 0000000000174fa4  /data/app/~~OdBMxcLjRbZre3jUFwqhtA==/com.zulip.flutter-Ivlz_Plsw-9nCNlt4y7GhA==/base.apk (io.flutter.embedding.android.h.K+344)
#22 pc 00000000005b98d4  /apex/com.android.art/lib64/libart.so (nterp_helper+3924) (BuildId: 7ece79c15d80914c83e60c9e93ac1684)
#23 pc 00000000001746a2  /data/app/~~OdBMxcLjRbZre3jUFwqhtA==/com.zulip.flutter-Ivlz_Plsw-9nCNlt4y7GhA==/base.apk (io.flutter.embedding.android.h.s+14)
#24 pc 00000000005b98d4  /apex/com.android.art/lib64/libart.so (nterp_helper+3924) (BuildId: 7ece79c15d80914c83e60c9e93ac1684)
#25 pc 00000000001759ee  /data/app/~~OdBMxcLjRbZre3jUFwqhtA==/com.zulip.flutter-Ivlz_Plsw-9nCNlt4y7GhA==/base.apk (io.flutter.embedding.android.g.onCreate+26)
#26 pc 00000000008e83ec  /data/misc/apexdata/com.android.art/dalvik-cache/arm64/boot.oat (android.app.Activity.performCreate+924)
#27 pc 000000000064b5a0  /data/misc/apexdata/com.android.art/dalvik-cache/arm64/boot.oat (android.app.Instrumentation.callActivityOnCreate+80)
#28 pc 000000000072cf94  /data/misc/apexdata/com.android.art/dalvik-cache/arm64/boot.oat (android.app.ActivityThread.performLaunchActivity+2708)
[… more frames …]
```

I didn't catch this earlier because it doesn't happen in debug mode.

Not sure what the issue is. Those top few frames sound like it's parsing some XML. Seems likely to be a bug in Flutter (or in Android) one way or another — in an APK, these XML files have been converted by the Android build system into its own binary format, so there really shouldn't be any parse errors after that stage.

In any case, we'll need to debug that before we can reland that change.

**Comment by gnprice** [comment]
@rajveermalviya, would you be up for investigating this after the small notification tweaks that are currently on your list?

**Comment by rajveermalviya** [comment]
> would you be up for investigating this after the small notification tweaks that are currently on your list?

@gnprice Sure, I'll self-assign it and take a at it look later.

**Comment by gnprice** [comment]
Here's an update on the state of this issue. Thanks to @rajveermalviya's investigation, we've learned:

* The strategy we'd been trying above unfortunately isn't going to work — the Flutter feature that we thought was there, isn't actually there.

  That strategy was to use an Android "network security config" XML file, just like the one we'd used on zulip-mobile; it looked like there was a Flutter feature to read that config and apply it to Dart's HTTP library. But that feature doesn't actually exist. The chat thread where we figured that out is [here](https://chat.zulip.org/#narrow/stream/243-mobile-team/topic/.23F461.20Crash.20when.20network.20policy.20supplied.20to.20flutter.20engine/near/1815549).

* There's an alternative strategy that should work: instead of the pure-Dart HTTP implementation that's in the Dart standard library, we could switch to `package:cronet_http` on Android. That uses Cronet, an HTTP client from Chromium. See [chat thread](https://chat.zulip.org/#narrow/stream/243-mobile-team/topic/Flutter.3A.20thoughts.20on.20native.20http.20libraries/near/1818003).

* Using Cronet (and perhaps `package:cupertino_http` for an analogous choice on iOS?) may be a good idea independent of this issue. But it's going to be nontrivial to do:
  * By default `package:cronet_http` depends on Google Play Services for the Cronet library it uses.
  * Some people use the Zulip app without Google Play Services, and we'd like it to keep working for them. There's a build-time flag we can use in order to bundle Cronet into the app instead.
  * But that bundled Cronet adds around 10 MB to the app's download size. That's a bigger size increase than I would feel great doing for all our users who don't need it (because they do have Play Services).
  * So we'd need to do something like build two different variants, with and without a bundled Cronet.

Given the greater complexity that it turns out will be involved in order to implement this, this isn't an issue we'll be able to get to right now. I still hope we'll be able to do this, but there are a number of other features we'll prioritize building first.

**Comment by rajveermalviya** [comment]
I think closing this was unintended, reopening.
(It happened because of the `... fix #461.` line in commit message.)

**Comment by gnprice** [comment]
Shoot, yes — thanks @rajveermalviya for catching that!

In that commit message I even used one of my usual circumlocutions for avoiding exactly this situation: "would be a fix for #nnn", instead of "would fix #nnn". But then a few sentences later when I wrote "a future effort to fix #461", I didn't notice that the magic phrase was appearing again and needed to be avoided again.

**Comment by gnprice** [comment]
More variations on keywords to help with finding this issue via GitHub search (since I was unable to do so when I tried a few minutes ago, and ended up switching to [a Zulip search](https://chat.zulip.org/#narrow/channel/48-mobile/search/certificate)): this is about the user specifying a custom CA to trust, for use with a TLS/SSL certificate.

**Comment by abunet** [comment]
Please reconsider my ticket #1990

Tested also on IOS phone, and the problem is identical. The only way to solve that i have found is to modify the codebase

**Comment by chrisbobbe** [comment]
#1990 was closed as a duplicate of this issue; it would be counterproductive to keep two open issues tracking the same thing. Please let us know if some part of #1990 isn't covered by this issue; see in particular the status update above: https://github.com/zulip/zulip-flutter/issues/461#issuecomment-2177588523

**Comment by abunet** [comment]
The code provided in #1990 solved the problem. The configuration that adds the user trust store doen't work. 
Please add an option like "skip ssl verify" when adding a new organization , and in the validate call back method, if the option Is checked, return true. Dirty but works on prem environments

  /// Construct an API connection that talks to a live Zulip server over the real network.
  ApiConnection.live({
    required Uri realmUrl,
    required int? zulipFeatureLevel, // required even though nullable; see field doc
    String? email,
    String? apiKey,
  }) : this(client: createCustomVerifierClient([realmUrl.host]),
            realmUrl: realmUrl, zulipFeatureLevel: zulipFeatureLevel,
            email: email, apiKey: apiKey, useBinding: true);
...
...
static IOClient createCustomVerifierClient(List<String> trustedHosts) {

    // The `trustedHosts` list is an external variable.
    // The callback function defined below will have access to this list.

    HttpClient httpClient = HttpClient()
      ..badCertificateCallback = (X509Certificate cert, String host, int port) {
// HERE YOU CAN RETRIEVE THE GLOBAL OPTION AND RETURN TRUE ONLY IF IT IS CHECKED
          return true;
      };

    return IOClient(httpClient);
  }

**Comment by gnprice** [comment]
> The code provided in [#1990](https://github.com/zulip/zulip-flutter/issues/1990) solved the problem.

The code you posted at https://github.com/zulip/zulip-flutter/issues/1990#issuecomment-3547806404 "solves" the "problem" by completely disabling validation of the server's certificate. That means that anyone on the network between your device and the server could impersonate your server using a made-up certificate, and the app would cheerfully send the impostor your credentials.

We're not going to ship an insecure "solution" like that.

I'm not excited about shipping an optional version of that "disable security" solution, either. I think a much better approach is the one originally requested in this thread, and also in #1990: accept a custom list of CAs to trust, so that the server's cert can be validated against those. That way the user's secrets remain protected against an attacker on the network.

The most likely strategy for that feature is the one described above at https://github.com/zulip/zulip-flutter/issues/461#issuecomment-2177588523 , swapping in a different HTTP client implementation like `package:cronet_http`.

An alternate approach would be to use the features our current HTTP client library `package:http` offers for specifying custom CAs to trust:
https://api.dart.dev/dart-io/SecurityContext-class.html
https://api.dart.dev/dart-io/HttpClient-class.html
One step there I don't know offhand how to do is to locate the right CA certificate files. I think that won't be something the Zulip core team finds time to take the lead on in the near term; but I'd be happy to review a [clean, well-presented PR](https://zulip.readthedocs.io/en/latest/contributing/reviewable-prs.html) that implemented that strategy.

**Comment by gnprice** [comment]
> One step there I don't know offhand how to do is to locate the right CA certificate files.

I took a few minutes just now to investigate this part. I think on at least iOS it looks like there isn't a stable, documented way to get this list. The most pertinent docs seem to be these:
https://developer.apple.com/documentation/security/configuring-a-trust
https://support.apple.com/en-us/103272

The first of those links to the other one like so:
> By default, the system trusts as an anchor only the root certificates packaged with the operating system (see the lists for [iOS](https://support.apple.com/en-us/HT204132) and [macOS](https://support.apple.com/en-us/HT202858)), but you can supplement this list.

But the latter just identifies specific certificates, by name, fingerprint, and other details, with different lists for different version of iOS. It doesn't give a way for software on a device to find the list that applies on that device, or the actual certificates in the list.

So I think on iOS the `package:http` approach with `SecurityContext` won't work, at least not with only documented APIs. (It's more likely to work on Android but I haven't checked there.)

**Comment by gnprice** [comment]
From a bit more research on the `package:cronet_http` option:
> Some people use the Zulip app without Google Play Services, and we'd like it to keep working for them. There's a build-time flag we can use in order to bundle Cronet into the app instead.

It looks like if we don't bundle the native Cronet library into the app, then on devices without Google Play Services, the package will still work; it'll just use a Java-based implementation that isn't as fast as the native one. That's my inference from the implementation, particularly around here:
https://source.chromium.org/chromium/chromium/src/+/5592ecea24eb0c34f6a5bf7aa33ac5683dd597e5:components/cronet/android/api/src/org/chromium/net/CronetProvider.java;bpv=1
(which I got to starting from here:
https://github.com/dart-lang/http/blob/f7fa470ee1280f26db337ddab5dc43665608a492/pkgs/cronet_http/lib/src/cronet_client.dart#L122-L134
)

I don't know how big that slowdown is, though. The native Cronet is likely faster than the Dart stdlib's HTTP implementation which we're using now, so the "slower" fallback Java version of Cronet might still be no slower than that.

So it's likely we could go ahead with `package:cronet_http` without the added build-time complexity of bundling the library into the app or of doing so conditionally. We'd need to validate that, though.

**Comment by abunet** [comment]
From my investigations, it seems that cronet_http does not support the management of custom CAs on iOS (on Android it works for both certificates installed on System and User Store).

I know the handling I proposed with badCertificateCallback was just a starting point. 

in the validation method instead of always returning true, you can check if the host name of the call corresponds the zulip organization url added in the gui.
Furthermore, as already mentioned, you can add an advanced option when adding a new organization where a "allows custom ca" check box. 
A second option could be to request, during the add new organization phase, to paste the base64 of the custom ca and verify it in the validationcallback.

As a developer, but at the same time as a user of Zulip, I prefer to be aware that I am using an unsafe mode, but still able to use the product, compared to the complete impossibility of using the mobile part

**Comment by gnprice** [comment]
Right, the plan would be to use `package:cronet_http` on Android, and [`package:cupertino_http`](https://pub.dev/packages/cupertino_http) on iOS. Because `package:cupertino_http` uses Apple's provided libraries for making an HTTP request, I expect it would use the system-configured custom CAs just like it uses system-configured VPNs and HTTP proxies.
